src/modules/sensors/vehicle_gps_position/VehicleGPSPosition.cpp

```diff
@@ -0,0 +1,742 @@
+		_NE_pos_offset_m[i](0) = constrain(_NE_pos_offset_m[i](0), -max_ne_offset(0), max_ne_offset(0));
+		_NE_pos_offset_m[i](1) = constrain(_NE_pos_offset_m[i](1), -max_ne_offset(1), max_ne_offset(1));
+		_hgt_offset_mm[i] = constrain(_hgt_offset_mm[i], -max_alt_offset, max_alt_offset);
+	}
+}
+
+void VehicleGPSPosition::apply_gps_offsets()
+{
+	// calculate offset corrected output for each physical GPS.
+	for (uint8_t i = 0; i < GPS_MAX_RECEIVERS; i++) {
+		// Add the sum of weighted offsets to the reference position to obtain the blended position
+		double lat_deg_now = (double)_gps_state[i].lat * 1.0e-7;
+		double lon_deg_now = (double)_gps_state[i].lon * 1.0e-7;
+		double lat_deg_res, lon_deg_res;
+		add_vector_to_global_position(lat_deg_now, lon_deg_now, _NE_pos_offset_m[i](0), _NE_pos_offset_m[i](1), &lat_deg_res,
+					      &lon_deg_res);
+		_gps_output[i].lat = (int32_t)(1.0E7 * lat_deg_res);
+		_gps_output[i].lon = (int32_t)(1.0E7 * lon_deg_res);
+		_gps_output[i].alt = _gps_state[i].alt + (int32_t)_hgt_offset_mm[i];
+
+		// other receiver data is used uncorrected
+		_gps_output[i].timestamp	= _gps_state[i].timestamp;
+		_gps_output[i].fix_type		= _gps_state[i].fix_type;
+		_gps_output[i].vel_m_s		= _gps_state[i].vel_m_s;
+		_gps_output[i].vel_n_m_s	= _gps_state[i].vel_n_m_s;
+		_gps_output[i].vel_e_m_s	= _gps_state[i].vel_e_m_s;
+		_gps_output[i].vel_d_m_s	= _gps_state[i].vel_d_m_s;
+		_gps_output[i].eph		= _gps_state[i].eph;
+		_gps_output[i].epv		= _gps_state[i].epv;
+		_gps_output[i].s_variance_m_s		= _gps_state[i].s_variance_m_s;
+		//_gps_output[i].gdop		= _gps_state[i].gdop; // TODO: add gdop
+		_gps_output[i].satellites_used		= _gps_state[i].satellites_used;
+		_gps_output[i].vel_ned_valid	= _gps_state[i].vel_ned_valid;
+		_gps_output[i].heading		= _gps_state[i].heading;
+		_gps_output[i].heading_offset	= _gps_state[i].heading_offset;
+	}
+}
+
+void VehicleGPSPosition::calc_gps_blend_output()
+{
+	// Convert each GPS position to a local NEU offset relative to the reference position
+	// which is defined as the positon of the blended solution calculated from non offset corrected data
+	Vector2f blended_NE_offset_m;
+	blended_NE_offset_m.zero();
+	float blended_alt_offset_mm = 0.0f;
+
+	for (uint8_t i = 0; i < GPS_MAX_RECEIVERS; i++) {
+		if (_blend_weights[i] > 0.0f) {
+			// calculate the horizontal offset
+			Vector2f horiz_offset{};
+			get_vector_to_next_waypoint((_gps_blended_state.lat / 1.0e7),
+						    (_gps_blended_state.lon / 1.0e7),
+						    (_gps_output[i].lat / 1.0e7),
+						    (_gps_output[i].lon / 1.0e7),
+						    &horiz_offset(0),
+						    &horiz_offset(1));
+
+			// sum weighted offsets
+			blended_NE_offset_m += horiz_offset * _blend_weights[i];
+
+			// calculate vertical offset
+			float vert_offset = (float)(_gps_output[i].alt - _gps_blended_state.alt);
+
+			// sum weighted offsets
+			blended_alt_offset_mm += vert_offset * _blend_weights[i];
+		}
+	}
```

Comment:
Are we not already doing the correction here: https://github.com/PX4/Firmware/pull/14447/files#diff-5c37c83ace9fdd92b81d362d265d2f21R503-R525 ?